python/ray/train/session.py

```diff
@@ -76,7 +76,7 @@ def node_rank(self) -> int:
     def get_dataset_shard(
         self,
         dataset_name: Optional[str] = None,
-    ) -> Optional[Union["Dataset", "DatasetPipeline"]]:
+    ) -> Optional["DatasetIterator"]:
```

Comment:
I would really like to keep backwards compatibility and a soft deprecation for at least 1 release since this API has been around for quite a while with multiple users using this.

We can introduce the following class in `ray.train._internal` temprorarily:
```
class DatasetIteratorWithBackwardCompat(DatasetIterator):
    def __getattr__(self, name):
        warnings.warn("session.get_dataset_shard no longer returns a dataset...")
        return getattr(self.base_dataset, name)
```

And have `session.get_dataset_shard` return this.

Reply:
Good call, thanks for the tip!